File: greis/catalog.py
"""What can be logged, and what each choice puts in the file.

One entry per GREIS message this application decodes. An entry names the
message, the columns it contributes to the CSV, and how often to ask the
receiver for it. The GUI renders this list; the command builder turns the
selection into ``em`` commands; the CSV writer builds its header from the
columns of whatever was selected. Adding a message means adding an entry
here and teaching the parser to decode it - nothing else changes.

Rates are **periods in seconds**, not frequencies, because that is what
GREIS's ``em`` command takes: ``em,,/msg/jps/PG:{0.01,0,0,0}`` asks for a
[PG] every 10 ms. The GUI shows the equivalent in Hz beside it, since a
period of 0.01 is easier to recognise as "100 Hz".
"""

from __future__ import annotations

from collections.abc import Callable
from dataclasses import dataclass

from greis.epoch import JavadEpoch


@dataclass(frozen=True)
class Column:
    """One CSV column: its header and how to get its value out of an epoch.

    ``value`` returns ``None`` for anything the receiver has not reported,
    which the writer renders as an empty cell. An empty cell and a zero are
    different facts and the file keeps them apart.
    """

    name: str
    value: Callable[[JavadEpoch], object | None]
    decimals: int | None = None
    """Decimal places for a float. ``None`` for values that are not floats
    (integers, text, timestamps), which are written as they are."""


@dataclass(frozen=True)
class LogMessage:
    """One selectable GREIS message."""

    code: str
    """The GREIS message id, e.g. ``"PG"``. Goes straight into the ``em``
    command path, so it is the message's real name and not a label."""
    label: str
    description: str
    columns: tuple[Column, ...]
    default_period_s: float
    mandatory: bool = False
    """[PG] is mandatory: it is what closes an epoch, so with it switched
    off the file would have no rows to put the other messages' values in.
    The GUI shows it checked and disabled rather than hiding it, so the
    reason is visible."""


def _fmt_date(epoch: JavadEpoch) -> str | None:
    return epoch.receiver_date.isoformat() if epoch.receiver_date is not None else None


def _fmt_time_of_day(epoch: JavadEpoch) -> str | None:
    """[ST]'s milliseconds-since-midnight as ``HH:MM:SS.mmm``.

    Rendered from the raw field rather than from ``utc_datetime`` so that
    [ST] on its own still produces a time - a log with the date switched
    off is a legitimate choice, not a broken one.
    """
    total_ms = epoch.time_of_day_ms
    if total_ms is None:
        return None
    ms = total_ms % 1000
    seconds = total_ms // 1000
    return f"{seconds // 3600:02d}:{seconds // 60 % 60:02d}:{seconds % 60:02d}.{ms:03d}"


def _fmt_utc_datetime(epoch: JavadEpoch) -> str | None:
    return epoch.utc_datetime.isoformat() if epoch.utc_datetime is not None else None


def _time_base(epoch: JavadEpoch) -> str | None:
    if epoch.time_base_is_utc is None:
        return None
    return "UTC" if epoch.time_base_is_utc else "GPS"


PG = LogMessage(
    code="PG",
    label="Position",
    description="Latitude, longitude, altitude, the receiver's own error estimate, and the solution type.",
    default_period_s=1.0,
    mandatory=True,
    columns=(
        # Nine decimals is about a tenth of a millimetre of latitude: past
        # anything a receiver can mean, and short of the point where the
        # binary double's own noise starts printing.
        Column("lat_deg", lambda e: e.latitude_deg, decimals=9),
        Column("lon_deg", lambda e: e.longitude_deg, decimals=9),
        Column("alt_m", lambda e: e.altitude_m, decimals=4),
        Column("pos_rms_m", lambda e: e.pos_rms_m, decimals=4),
        Column("sol_type", lambda e: e.sol_type),
        Column("sol_type_label", lambda e: e.sol_type_label),
    ),
)

VG = LogMessage(
    code="VG",
    label="Velocity",
    description="North/east/up velocity components with the receiver's error estimate, plus ground and 3D speed.",
    default_period_s=1.0,
    columns=(
        Column("vel_north_mps", lambda e: e.vel_north_mps, decimals=4),
        Column("vel_east_mps", lambda e: e.vel_east_mps, decimals=4),
        Column("vel_up_mps", lambda e: e.vel_up_mps, decimals=4),
        Column("vel_rms_mps", lambda e: e.vel_rms_mps, decimals=4),
        Column("vel_ground_mps", lambda e: e.vel_ground_mps, decimals=4),
        Column("vel_3d_mps", lambda e: e.vel_3d_mps, decimals=4),
    ),
)

ST = LogMessage(
    code="ST",
    label="Time of day",
    description="The receiver's clock, as milliseconds since midnight on its own time base.",
    default_period_s=1.0,
    columns=(Column("rx_time_of_day", _fmt_time_of_day),),
)

RD = LogMessage(
    code="RD",
    label="Date",
    description=(
        "The receiver's date and which time base it is on. Combined with the time of day "
        "into a full UTC timestamp, so the timestamp column stays empty unless Time of day "
        "is selected too."
    ),
    default_period_s=1.0,
    columns=(
        Column("rx_date", _fmt_date),
        Column("rx_time_base", _time_base),
        Column("rx_datetime_utc", _fmt_utc_datetime),
    ),
)

NP = LogMessage(
    code="NP",
    label="Satellites",
    description="How many satellites of each constellation went into the solution.",
    default_period_s=1.0,
    columns=(
        Column("sv_gps", lambda e: e.sv_gps),
        Column("sv_glonass", lambda e: e.sv_glonass),
        Column("sv_galileo", lambda e: e.sv_galileo),
        Column("sv_beidou", lambda e: e.sv_beidou),
        Column("sv_total", lambda e: e.sv_total),
    ),
)

CATALOG: tuple[LogMessage, ...] = (PG, VG, ST, RD, NP)
"""In the order they are offered, which is the order their columns appear
in the file: where you are, how fast, when, and with what."""

BY_CODE: dict[str, LogMessage] = {message.code: message for message in CATALOG}

HOST_TIME_COLUMN = Column("host_time_utc", lambda e: e.received_at.isoformat())
"""Always the first column, whatever is selected. It is the one timestamp
that exists for every row even before the receiver's clock is known, so a
file is never without a time axis."""

PERIOD_CHOICES_S: tuple[float, ...] = (0.01, 0.02, 0.05, 0.1, 0.2, 0.5, 1.0, 2.0, 5.0, 10.0, 30.0, 60.0)
"""The periods offered in the GUI. 0.01 s is the fastest javad-udp-target
drives a Delta at over a serial link and the fastest worth offering: below
that the messages outrun the port before they outrun the receiver."""


def period_label(period_s: float) -> str:
    """``0.01`` reads as ``10 ms (100 Hz)``; ``30`` reads as ``30 s``.

    Sub-second periods are named in milliseconds and in hertz because that
    is how a rate is usually asked for; a period of seconds is already
    clear on its own.
    """
    if period_s < 1.0:
        milliseconds = period_s * 1000
        millisecond_text = f"{milliseconds:.0f}" if milliseconds.is_integer() else f"{milliseconds:g}"
        return f"{millisecond_text} ms ({1 / period_s:g} Hz)"
    seconds_text = f"{period_s:.0f}" if float(period_s).is_integer() else f"{period_s:g}"
    return f"{seconds_text} s"

File: ios/GreisKit/Sources/GreisKit/Catalog.swift
import Foundation

/// What can be logged, and what each choice puts in the file.
/// Ported from `greis/catalog.py`.
///
/// One entry per GREIS message this application decodes. Adding a message
/// means adding an entry here and teaching the parser to decode it —
/// nothing else changes, because the screen renders this list, the command
/// builder turns the selection into `em` commands, and the CSV writer
/// builds its header from the columns of whatever was selected.
public struct LogColumn: Sendable {
    public let name: String
    /// How many decimals for a floating-point value. `nil` for values that
    /// are not floats — integers, labels, timestamps — which are written as
    /// they are.
    public let decimals: Int?
    public let value: @Sendable (JavadEpoch) -> CellValue

    public init(_ name: String, decimals: Int? = nil, value: @escaping @Sendable (JavadEpoch) -> CellValue) {
        self.name = name
        self.decimals = decimals
        self.value = value
    }
}

/// One cell. `.missing` is written as an empty cell and never as a zero:
/// "the receiver did not report an altitude" and "the receiver reported an
/// altitude of zero" are different facts, and a reader averaging a column
/// has to be able to tell them apart.
public enum CellValue: Sendable, Equatable {
    case missing
    case number(Double)
    case integer(Int)
    case text(String)

    init(_ value: Double?) { self = value.map { .number($0) } ?? .missing }
    init(_ value: Int?) { self = value.map { .integer($0) } ?? .missing }
    init(_ value: String?) { self = value.map { .text($0) } ?? .missing }
}

public struct LogMessage: Identifiable, Sendable {
    /// The GREIS message id, e.g. "PG". It goes straight into the `em`
    /// command path, so it is the message's real name and not a label.
    public let code: String
    public let label: String
    public let detail: String
    public let columns: [LogColumn]
    public let defaultPeriod: Double
    /// [PG] is mandatory: it is what closes an epoch, so with it switched
    /// off the file would have no rows to put the other messages' values in.
    /// The screen shows it on and disabled rather than hiding it, so the
    /// reason is visible instead of mysterious.
    public let mandatory: Bool
    /// Computed here from what another message already carries, rather than
    /// asked for. A derived entry gets no `em` command and has no rate — it
    /// arrives exactly as often as the message it is computed from — so the
    /// command builder skips it and the screen hides its rate.
    public var derived: Bool = false
    /// Asked for on a timer with `print`, rather than subscribed to with
    /// `em`. GREIS has no way to stream a parameter-tree value the way it
    /// streams a message, so — like a derived entry — this gets no `em`
    /// command. Unlike a derived one, though, something still has to go
    /// out over the wire on a schedule to keep it current; `defaultPeriod`
    /// is that schedule's period rather than a message rate.
    public var polled: Bool = false

    public var id: String { code }
}

public enum Catalog {

    public static let hostTimeColumn = LogColumn("host_time_utc") { epoch in
        .text(ISO8601.string(from: epoch.receivedAt))
    }

    public static let pg = LogMessage(
        code: "PG",
        label: "Position",
        detail: "Latitude, longitude, altitude, the receiver's own error estimate, and the solution type.",
        columns: [
            // Nine decimals is about a tenth of a millimetre of latitude:
            // past anything a receiver can mean, and short of the point
            // where the binary double's own noise starts printing.
            LogColumn("lat_deg", decimals: 9) { .init($0.latitudeDeg) },
            LogColumn("lon_deg", decimals: 9) { .init($0.longitudeDeg) },
            LogColumn("alt_m", decimals: 4) { .init($0.altitudeM) },
            LogColumn("pos_rms_m", decimals: 4) { .init($0.posRmsM) },
            LogColumn("sol_type") { .init($0.solType) },
            LogColumn("sol_type_label") { epoch in
                epoch.solType == nil ? .missing : .text(epoch.solTypeLabel)
            },
        ],
        defaultPeriod: 1.0,
        mandatory: true
    )

    public static let vg = LogMessage(
        code: "VG",
        label: "Velocity",
        detail: "North/east/up velocity components with the receiver's error estimate, plus ground and 3D speed.",
        columns: [
            LogColumn("vel_north_mps", decimals: 4) { .init($0.velNorthMps) },
            LogColumn("vel_east_mps", decimals: 4) { .init($0.velEastMps) },
            LogColumn("vel_up_mps", decimals: 4) { .init($0.velUpMps) },
            LogColumn("vel_rms_mps", decimals: 4) { .init($0.velRmsMps) },
            LogColumn("vel_ground_mps", decimals: 4) { .init($0.velGroundMps) },
            LogColumn("vel_3d_mps", decimals: 4) { .init($0.vel3DMps) },
        ],
        defaultPeriod: 1.0,
        mandatory: false
    )

    public static let st = LogMessage(
        code: "ST",
        label: "Time of day",
        detail: "The receiver's clock, as milliseconds since midnight on its own time base.",
        columns: [
            // Rendered from the raw field rather than from utcDatetime, so
            // that [ST] on its own still produces a time: a log with the
            // date switched off is a legitimate choice, not a broken one.
            LogColumn("rx_time_of_day") { epoch in
                guard let total = epoch.timeOfDayMs else { return .missing }
                let ms = total % 1000
                let seconds = total / 1000
                return .text(String(format: "%02d:%02d:%02d.%03d",
                                    seconds / 3600, (seconds / 60) % 60, seconds % 60, ms))
            }
        ],
        defaultPeriod: 1.0,
        mandatory: false
    )

    public static let rd = LogMessage(
        code: "RD",
        label: "Date",
        detail: "The receiver's date and which time base it is on. Combined with the time of day into a full UTC timestamp, so that column stays empty unless Time of day is selected as well.",
        columns: [
            LogColumn("rx_date") { .init($0.receiverDateISO) },
            LogColumn("rx_time_base") { epoch in
                guard let isUTC = epoch.timeBaseIsUTC else { return .missing }
                return .text(isUTC ? "UTC" : "GPS")
            },
            LogColumn("rx_datetime_utc") { epoch in
                guard let stamp = epoch.utcDatetime else { return .missing }
                return .text(ISO8601.string(from: stamp))
            },
        ],
        defaultPeriod: 1.0,
        mandatory: false
    )

    public static let np = LogMessage(
        code: "NP",
        label: "Satellites",
        detail: "How many satellites of each constellation went into the solution.",
        columns: [
            LogColumn("sv_gps") { .init($0.svGPS) },
            LogColumn("sv_glonass") { .init($0.svGLONASS) },
            LogColumn("sv_galileo") { .init($0.svGalileo) },
            LogColumn("sv_beidou") { .init($0.svBeiDou) },
            LogColumn("sv_total") { .init($0.svTotal) },
        ],
        defaultPeriod: 1.0,
        mandatory: false
    )

    public static let jstar = LogMessage(
        code: "JSTAR",
        label: "J-Star lock",
        detail: "Whether the receiver has locked onto a JAVAD J-Star L-Band correction beam, and which one. Read from the receiver's own parameters on a timer rather than a subscribed message, because GREIS has no message for it.",
        columns: [
            LogColumn("jstar_beam_name") { .init($0.jstarBeamName) },
            LogColumn("jstar_snr") { .init($0.jstarSNR) },
            LogColumn("jstar_locked") { epoch in
                guard let locked = epoch.jstarLocked else { return .missing }
                return .text(locked ? "True" : "False")
            },
        ],
        defaultPeriod: 2.0,
        mandatory: false,
        polled: true
    )

    public static let radians = LogMessage(
        code: "RAD",
        label: "Position in radians",
        detail: "Latitude and longitude in radians, the unit [PG] sends them in before the parser converts once to degrees. Altitude has no radian form: a height is a length, not an angle.",
        columns: [
            // Twelve decimals of a radian is 1e-12 rad, about a hundredth
            // of a millimetre on the ground — finer than the nine decimals
            // of a degree beside it, so this is never the column that
            // rounds.
            LogColumn("lat_rad", decimals: 12) { .init($0.latitudeRad) },
            LogColumn("lon_rad", decimals: 12) { .init($0.longitudeRad) },
        ],
        defaultPeriod: 1.0,
        mandatory: false,
        derived: true
    )

    public static let ecef = LogMessage(
        code: "ECEF",
        label: "ECEF position",
        detail: "The same position as X, Y and Z in metres from the centre of the earth, computed from [PG] on the WGS-84 ellipsoid. Costs the receiver nothing.",
        columns: [
            // Four decimals is a tenth of a millimetre, matching the
            // altitude it is computed from. A double carries eleven
            // significant digits at 6 400 km without strain, so the
            // precision of the input survives the transform.
            LogColumn("ecef_x_m", decimals: 4) { .init($0.ecefXM) },
            LogColumn("ecef_y_m", decimals: 4) { .init($0.ecefYM) },
            LogColumn("ecef_z_m", decimals: 4) { .init($0.ecefZM) },
        ],
        defaultPeriod: 1.0,
        mandatory: false,
        derived: true
    )

    /// In the order they are offered, which is the order their columns
    /// appear in the file: where you are, how fast, when, and with what.
    public static let all: [LogMessage] = [pg, vg, st, rd, np, jstar, radians, ecef]

    public static func message(code: String) -> LogMessage? {
        all.first { $0.code == code }
    }

    /// The periods offered. 0.01 s is the fastest javad-udp-target drives a
    /// Delta and the fastest worth offering.
    public static let periods: [Double] = [0.01, 0.02, 0.05, 0.1, 0.2, 0.5, 1, 2, 5, 10, 30, 60]

    /// `0.01` reads as "10 ms (100 Hz)"; `30` reads as "30 s".
    public static func periodLabel(_ period: Double) -> String {
        if period < 1 {
            let ms = period * 1000
            let msText = ms == ms.rounded() ? String(Int(ms)) : String(ms)
            let hz = 1 / period
            let hzText = hz == hz.rounded() ? String(Int(hz)) : String(format: "%g", hz)
            return "\(msText) ms (\(hzText) Hz)"
        }
        return period == period.rounded() ? "\(Int(period)) s" : String(format: "%g s", period)
    }
}
